### ISSUE TITLE
Kubelet should not add IPv6 entries to /etc/hosts on systems where IPv6 is disabled

### ISSUE THREAD

**Issue by PaulFurtado** [kubernetes/kubernetes#124686]
### What happened?

Kubelet unconditionally adds IPv6 entries to `/etc/hosts` even when system has IPv6 totally disabled. This results in strange errors due to "happy eyeballs" fallback behavior: when clients see both an IPv4 and an IPv6 address for a host, they first try IPv4 and then fall back to IPv6, but if IPv6 is disabled in the kernel, they then get an "Address family not supported by protocol" exception instead of "Connection refused".

This happens with standard HTTP clients, etc, but here's an example with netcat:
```
# nc -v 127.0.0.1 1111
Ncat: Version 7.92 ( https://nmap.org/ncat )
Ncat: Connection refused.
```
vs
```
Ncat: Version 7.92 ( https://nmap.org/ncat )
Ncat: Connection to 127.0.0.1 failed: Connection refused.
Ncat: Trying next address...
libnsock nsock_make_socket(): Socket trouble: Address family not supported by protocol
Ncat: Address family not supported by protocol.
```

That results in clients needing to handle a different exception. However, even if IPv6 were enabled on the system, it would still be invalid to put loopback addresses in `/etc/hosts` for localhost when they are not actually configured on the loopback interface.

Code here: https://github.com/kubernetes/kubernetes/blob/v1.29.4/pkg/kubelet/kubelet_pods.go#L373-L377

### What did you expect to happen?

Kubelet should not add localhost entries to `/etc/hosts` for addresses that do not actually exist on the pod's loopback interface.

### How can we reproduce it (as minimally and precisely as possible)?

I don't think you need a reproducer for this one since the code is just unconditionally doing this. However, if you're looking to test this on a system with ipv6 disabled such that you get the "Address family not supported by protocol" error, or guarantee that CNI does not add an IPv6 loopback address, you can boot your kernel with `ipv6.disable=1` on the kernel command line.

### Kubernetes version

N/A all versions impacted

### Cloud provider

N/A

### OS version

N/A

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

**Comment by aojea** [comment]
interesting behavior, cc: @thockin @danwinship , 🤔 this behavior is at least 8 years old  https://github.com/kubernetes/kubernetes/commit/b1f91bd510b680e7d455623225ca2da5804ad3e2
Should we make optional adding the ipv6 localhost addresses?
If optional, should it be automatically detected or user configured?
...

**Comment by danwinship** [comment]
> This happens with standard HTTP clients, etc, but here's an example with netcat:

I assume you're connecting to `localhost` in the second example? (You didn't include the command line in the paste.)

> this behavior is at least 8 years old [b1f91bd](https://github.com/kubernetes/kubernetes/commit/b1f91bd510b680e7d455623225ca2da5804ad3e2)

That commit was just moving code around; this goes back to the original implementation of `/etc/hosts` in k8s (#16052) which seems to have mostly just copied docker's behavior. (Or at least, it copied what the author believed was docker's behavior, but possibly docker wouldn't actually have put `::1` there if ipv6 was disabled?)

Anyway, yes, we should not be adding an entry for `::1` if the loopback interface inside the pod does not actually get assigned that IP. (It is probably better for us to actually check the interface rather than to try to guess whether it will get that IP based on sysctls.)

(And we probably shouldn't be adding those [weird IPv6 multicast aliases](https://github.com/kubernetes/kubernetes/blob/v1.29.4/pkg/kubelet/kubelet_pods.go#L374-L377) under any circumstances, though getting rid of them now would probably require a feature gate since someone might be depending on them...)

**Comment by HirazawaUi** [comment]
> (And we probably shouldn't be adding those [weird IPv6 multicast aliases](https://github.com/kubernetes/kubernetes/blob/v1.29.4/pkg/kubelet/kubelet_pods.go#L374-L377) under any circumstances, though getting rid of them now would probably require a feature gate since someone might be depending on them...)

For this  feature gate should we use the recently published guide on [skipping Alpha for feature gated changes](https://github.com/kubernetes/community/pull/7828)? it's simple enough and the risk is low enough.

**Comment by danwinship** [comment]
I'd probably use a [`Deprecated` gate](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#deprecation), so it's even fewer stages. (`featuregate.Deprecated` → wait a few releases → delete stuff)

**Comment by HirazawaUi** [comment]
Thank you for your suggestion, I will implement it in two PRs respectively

1. By checking whether the loopback interface is assigned `::1` IP to deciding whether to add an entry for `::1` in `/etc/hosts`.
2. Add a new `Deprecated feature gate` to prevent weird IPv6 multicast aliases from being added to the pod.

/assign

**Comment by thockin** [comment]
Yeah, this was probably a snaphsot of what docker was doing, but we took ownership of the file so docker would stop breaking it (https://github.com/kubernetes/kubernetes/issues/14633).

**Comment by uablrek** [comment]
Is this issue sig/network or sig/node? Or both?

**Comment by uablrek** [comment]
> If optional, should it be automatically detected or user configured?

I am usually not in favor for any "automatically detected" things, but in this case I think we should check the node `lo` interface and do the same. And, since I want to promote IPv6, don't forget the opposite:

Kubelet should not add IPv4 entries if IPv4 is disabled

**Comment by HirazawaUi** [comment]
> Is this issue sig/network or sig/node? Or both?

I believe it's both.
/sig node



> Kubelet should not add IPv4 entries if IPv4 is disabled

Good point, I will add it to #124758.

**Comment by thockin** [comment]
Is this a case of the kernel having IPv6 totally compiled out, or disabled in some software way?  I cannot reproduce the symptom.

**Comment by HirazawaUi** [comment]
I can't reproduce it either.
```
➜  ~ sysctl net.ipv6.conf.all.disable_ipv6
net.ipv6.conf.all.disable_ipv6 = 1
➜  ~
➜  ~ nc -v 127.0.0.1 222
nc: connect to 127.0.0.1 port 222 (tcp) failed: Connection refused
➜  ~
➜  ~ nc -v ip6-localhost 443
nc: connect to ip6-localhost (::1) port 443 (tcp) failed: Cannot assign requested address
```

Considering that nsock_make_socket() is a function in the nmap project,
ref: https://github.com/nmap/nmap/blob/0e3a51c992a0ac9c189f44413fdf7c2440cdaa05/nsock/src/nsock_connect.c#L139-L167

Is this issue present only in specific versions of Ncat?
